File: Python/162-findPeakElement.py
class Solution:
    # @param nums, an integer[]
    # @return an integer
    def findPeakElement(self, nums):
        # length = len(nums)
        # if length == 1: return 0
        return self.searchPeak(nums, 0, len(nums) - 1)
    
    def searchPeak(self, nums, start, end):
        if end == start:
           return start
        if end == start + 1:
           return start if nums[start] > nums[end] else end
        mid = (end + start) / 2
        if nums[mid] > nums[mid + 1]:
           return self.searchPeak(nums, start, mid)
        elif nums[mid] < nums[mid + 1]:
             return self.searchPeak(nums, mid, end)

if __name__ == '__main__':
   s = Solution()
   print s.findPeakElement([1, 5, 2, 3, 4, 2, 1])
   print s.findPeakElement([2, 1])
   print s.findPeakElement([1])

File: C++/162-findPeakElement.cpp
#include "include/common.h"

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int start = 0, end = nums.size() - 1, mid;
        if (nums.size() < 2) {
            return 0;
        };
        while (start < end) {
               // if (start == end) {
               //     return start;
               // }
               if (end == start + 1) {
                   return nums[end] > nums[start] ? end : start;
               };
               mid = (start + end) / 2;
               if (nums[mid] > nums[mid + 1]) {
                   end = mid;
               }
               else if (nums[mid] < nums[mid + 1]) {
                        start = mid;
               }
               else {
                   start = mid;
               };
        };
        return -1;
    }
};

int main(void) {
    Solution s;
    Utils u;
    // int iarr [] = {1, 2, 3, 1};
    int iarr [] = {1, 2, 3};
    // int iarr [] = {1, 2, 1, 3, 4, 5, 7, 6};
    vector<int> nums(iarr, iarr + sizeof(iarr) / sizeof(iarr[0]));
    cout << s.findPeakElement(nums) << endl;
    return 0;
}